src/app/AttributePathExpandIterator.h

```diff
@@ -0,0 +1,127 @@
+#include <app/EventManagement.h>
+#include <app/InteractionModelDelegate.h>
+#include <lib/core/CHIPCore.h>
+#include <lib/core/CHIPTLVDebug.hpp>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/DLLUtil.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <messaging/Flags.h>
+#include <protocols/Protocols.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * AttributePathExpandIterator is used to iterate over a linked list of ClusterInfo-s.
+ * The AttributePathExpandIterator is copiable, however, the given cluster info must be valid when calling Next().
+ *
+ * AttributePathExpandIterator will expand attribute paths with wildcards, and only emit exist paths for ClusterInfo with
+ * wildcards. For ClusterInfo with a concrete path (i.e. does not contain wildcards), AttributePathExpandIterator will emit them
+ * as-is.
+ *
+ * The typical use of AttributePathExpandIterator may look like:
+ * ConcreteAttributePath path;
+ * for (AttributePathExpandIterator iterator(clusterInfo); iterator.Get(path); iterator.Next()) {...}
+ *
+ * The iterator does not copy the given ClusterInfo, The given ClusterInfo must be valid when using the iterator.
+ * If there are new endpoints, clusters or attributes adds, AttributePathExpandIterator must be reinitialized.
+ *
+ * A initialized iterator will return the first valid path, no need to call Next() before calling Get() for the first time.
+ *
+ * Note: The Next() and Get() are two separate operations by design since a possible call of this iterator might be:
+ * - Get()
+ * - Chunk full, return
+ * - In a new chunk, Get()
+ *
+ * TODO: The ClusterInfo may support a group id, the iterator should be able to call group data provider to expand the group id.
+ */
+class AttributePathExpandIterator
+{
+public:
+    AttributePathExpandIterator(ClusterInfo * aClusterInfo);
+
+    /**
+     * Proceed the iterator to the next attribute path in the given cluster info.
+     *
+     * Feturns false if AttributePathExpandIterator has exhausted all paths in the given ClusterInfo list.
+     */
+    bool Next();
+
+    /**
+     * Fills the aPath with the path the iterator currently points to.
+     * Returns false if the iterator is not pointing to a valid path (i.e. it has exhausted the cluster info).
+     */
+    bool Get(ConcreteAttributePath & aPath)
+    {
+        aPath = mOutputPath;
+        return Valid();
+    }
+
+    /**
+     * Returns if the iterator is valid (not exhausted). An iterator is exhausted if and only if:
+     * - Next() is called after iterating last path.
+     * - Iterator is initialized wirh a null ClusterInfo.
+     */
+    inline bool Valid() const
+    {
+        // Check if we can emitting a valid path, we are not checking mpClusterInfo since we may just visited a ClusterInfo with
+        // concrete attribute path, and it is the last one in the linked list.
```

Comment:
```suggestion
        // Check whether we can emit a valid path.  We are not checking mpClusterInfo since we may have just visited a ClusterInfo with
        // a concrete attribute path, and it is the last one in the linked list.
```